### PR TITLE
feat(router): add skipNavigation option to routerLink

### LINKED ISSUE #13980: routerLink and `null` value should remove `href` attribute from link
```
[ ] bug report
[ x ] feature request
[ ] support request
```

**What is the motivation / use case for changing the behavior?**
Currently null or empty value passed to `routerLink` adds current page `href` value to link that is not useful. There is a lot of usecases for 'disabled' links that should not navigate anywhere and should not has `href` attribute at all.
It would be useful to have null/empty value support by `routerLink` that will results with no `href` attribute on link at all to prevent any navigation/link copying.

### LINKED ISSUE #31154: Add 'skipNavigation' boolean to routerLink to disable navigating
# 🚀 feature request

### Relevant Package
This feature request is for `@angular/router` (specifically the `routerLink` directive)

### Description
At the moment it doesn't seem possible to combine the power of `routerLinkActive` (for class changes) with a `routerLink` that doesn't navigate.  
This can be useful for instance when you have a navigation of multiple levels, and only the child levels are allowed to actually navigate (but you still want to leverage the `routerLinkActive` directive to add/remove classes on the links that shouldn't navigate.

### Describe the solution you'd like
* Add `skipNavigation` to the `routerLink` directive to block the actual navigation of the `onClick` handler.

``` typescript
  @HostListener('click', ['$event.button', '$event.ctrlKey', '$event.metaKey', '$event.shiftKey'])
  onClick(button: number, ctrlKey: boolean, metaKey: boolean, shiftKey: boolean): boolean {
    if (button !== 0 || ctrlKey || metaKey || shiftKey) {
      return true;
    }

    if (typeof this.target === 'string' && this.target !== '_self') {
      return true;
    }

    const extras = {
      skipLocationChange: attrBoolValue(this.skipLocationChange),
      replaceUrl: attrBoolValue(this.replaceUrl),
      state: this.state
    };
+    if(!attrBoolValue(this.skipNavigation)) {
+      this.router.navigateByUrl(this.urlTree, extras);
+    }
    return false;
  }
```

### Describe alternatives you've considered
* making the `routerLink` of the non-navigating links empty (or /) doesn't help, since `routerLinkActive` will match all of them when using `{ exact: false }`.
* using `{exact: true}` works for not matching all the root levels, but this will in turn never actually show a clear view of what links are currently active based on the URL (current link + all it's parents should be set to active, not only the clicked link).
* good practice would be using a guard, but for this specific situation it seems to be quite the overkill when it can just be fixed with a boolean that is false by default (so doesn't impact anything but the actual case).

### Related Issues
#10431
#22588
#13980

### PR DESCRIPTION
**Pull request by jordanpowell88**
The RouterLink directive now has a 'skipNavigation' boolean which disables navigating.

Closes #31154
Closes #13980

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [X] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [X ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #31154

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

### CODE REVIEW

**Comment by jordanpowell88** [comment]
@googlebot I signed it!

File: packages/router/src/directives/router_link.ts [hunk 1]
```diff
@@ -160,6 +160,7 @@ export class RouterLink implements OnChanges {
    * @see {@link NavigationBehaviorOptions#replaceUrl NavigationBehaviorOptions#replaceUrl}
    * @see {@link Router#navigateByUrl Router#navigateByUrl}
    */
+  @Input() disableNavigation?: boolean;
```
Comment (atscott) [inline packages/router/src/directives/router_link.ts:163]: I think there was a mistake in the rebase - this option now has the documentation of `replaceurl`. Please move it below `replaceUrl` and add documentation for it.

The same goes for below in `RouterLinkWithHref`

**Review by atscott — CHANGES_REQUESTED** [review]
Hi @jordanpowell88, thanks for the contribution. A couple quick things in addition to the comment on the code:

* Could you please fix the lint and tests errors from CI?
* Please update your commit message to include an explanation of why this option is useful
* Please update your commit message to include "fixes #31154" so it's automatically closed when this is merged.

**Comment by jordanpowell88** [comment]
@atscott when I run tests locally on upstream/master I'm getting failures? What am I missing?

**Comment by atscott** [comment]
@jordanpowell88 - I'm not sure exactly what tests are failing for you locally. For the CI failure, all you would need to do to fix it would be  `yarn bazel run //packages/router:router_api.accept`. This will change the router api golden file, which you should add to the commit. The lint failure is because your commit message doesn't match our guidelines, but would be addressed if you edit it based on my previous comment. 

If you do those two things, the tests will at least pass on CI. Hope that helps!

**Comment by jordanpowell88** [comment]
> @jordanpowell88 - I'm not sure exactly what tests are failing for you locally. For the CI failure, all you would need to do to fix it would be `yarn bazel run //packages/router:router_api.accept`. This will change the router api golden file, which you should add to the commit. The lint failure is because your commit message doesn't match our guidelines, but would be addressed if you edit it based on my previous comment.
> 
> If you do those two things, the tests will at least pass on CI. Hope that helps!

Yes that was helpful. Let me give it another shot. Thanks again as this is my first contribution to the Angular project

File: packages/router/src/directives/router_link.ts [hunk 2]
```diff
@@ -295,6 +306,13 @@ export class RouterLinkWithHref implements OnChanges, OnDestroy {
    * @see {@link NavigationBehaviorOptions#state NavigationBehaviorOptions#state}
    * @see {@link Router#navigateByUrl Router#navigateByUrl}
    */
+  @Input() disableNavigation?: boolean;
```
Comment (atscott) [inline packages/router/src/directives/router_link.ts:309]: Please fix the documentation for both `disableNavigation` and `state`. You can see them in the preview here: https://pr39916-1504cef.ngbuilds.io/api/router/RouterLink.

File: packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_directives/directives/matching/factory.js [hunk 3]
```diff
@@ -0,0 +1,3 @@
+MyComponent.ɵfac = function MyComponent_Factory(t) {
+  return new (t || MyComponent)();
+};
```
Comment (JoostK) [inline packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_directives/directives/matching/factory.js:3]: This looks like a rebase issue; I believe this was has been removed on master.

Comment (jordanpowell88) [reply packages/compiler-cli/test/compliance/test_cases/r3_view_compiler_directives/directives/matching/factory.js]: should be removed now. Thanks for catching

File: packages/router/src/directives/router_link.ts [hunk 4]
```diff
@@ -7,7 +7,7 @@
  */
 
 import {LocationStrategy} from '@angular/common';
-import {Attribute, Directive, ElementRef, HostBinding, HostListener, Input, OnChanges, OnDestroy, Renderer2, SimpleChanges} from '@angular/core';
+import {Attribute, Directive, ElementRef, HostBinding, HostListener, Input, OnChanges, OnDestroy, Renderer2, SimpleChanges,} from '@angular/core';
```
Comment (atscott) [inline packages/router/src/directives/router_link.ts:10]:
```suggestion
import {Attribute, Directive, ElementRef, HostBinding, HostListener, Input, OnChanges, OnDestroy, Renderer2, SimpleChanges} from '@angular/core';
```
Please limit the changes in this PR to the new feature only. If you want to add commas for formatting, this should be done in a separate PR.

**Comment by atscott** [comment]
@jordanpowell88 are you still working on this? I see there are still issues with the tests and documentation, but don't want to keep commenting on every commit until you're ready for re-review. Please use the `re-request review` button in the GitHub UI when you're ready and/or tag me in a comment. Thanks.

**Comment by petebacondarwin** [comment]
@atscott - shall we just close this PR as it has not been updated since before the New Year.

**Comment by atscott** [comment]
@petebacondarwin - sounds good. Closing due to lack of activity. We still need to decide if we want to handle disabling a `routerlink` to close out the linked issues and how we would go about doing that. There are several of them and there needs to be a cohesive solution that takes accessibility into account.
